### ISSUE TITLE
Why does Alacritty terminal gets more attention than Kitty?

### ISSUE THREAD

**Issue by horta** [kovidgoyal/kitty#2701]
Kitty is super fast. Very stable. Well maintained. Provide the minimal features (split screen, tabs) for a fully working terminal (no need to add another indirection layer, like tmux, that will add complexity to your workflow, or hope that your OS have a window manager that will fill the features gap left by Alacritty).

I understand the point of Alacritty, their reasons, and it is a great terminal but it seems that Kitty has all of Alacritty's strenghts plus do not lack the minimal set of features to make it a fully working terminal out-of-the-box.

**Comment by SolitudeSF** [comment]
it doesnt have rust circlejerk

**Comment by kovidgoyal** [comment]
Well, go out and spread the word :)

**Comment by Luflosi** [comment]
Alacritty runs on Windows and is the default terminal emulator on Sway.

**Comment by kovidgoyal** [comment]
Closing as not really a bug, but feel free to continue the discussion.

**Comment by ctrlcctrlv** [comment]
Alacritty is a gimmick TE. That's why. Its authors are good at marketing. Its main gimmick is its alleged speed. Meanwhile we're really trying to build a terminal emulator useful for most people, we don't need to make ridiculous classrooms, which, if true, are only so due to severe lack of useful features.

I wrote a blog post a while ago about why I don't like Alacritty or indeed most of the Rust ecosystem: https://gist.github.com/ctrlcctrlv/978b3ee4f55d4b4ec415a985e01cb1c9

**Comment by kovidgoyal** [comment]
And here we have some hardware based measurements showing kitty's latency is optimal, and much better than alacritty's: https://thume.ca/2020/05/20/making-a-latency-tester/

**Comment by ctrlcctrlv** [comment]
That doesn't surprise me at all @kovidgoyal. The Alacritty devs have a very lopsided idea of what kind of speed even matters. They think it's just CPU operations...lol.

**Comment by pentago** [comment]
To those who land on this discussion..
The main difference today (Oct 2020) is:
Kitty doesn't work on MacOS High Sierra (perhaps even some builds up) while Allacrity does.

https://i.imgur.com/A3oICfi.png

**Comment by ctrlcctrlv** [comment]
@pentago start a new issue.

**Comment by Luflosi** [comment]
This is actually known: https://github.com/kovidgoyal/kitty/issues/2783.
Apparently the notarisation broke it though there is probably a workaround that we haven't figured out.

**Comment by pentago** [comment]
@ctrlcctrlv no need for new issue, I'm happy Alacritty user as compared to Kitty, it works on my antique of an OS.

**Comment by ctrlcctrlv** [comment]
So you're just here to troll then and not improve Kitty?

**Comment by CaninoDev** [comment]
> Kitty is super fast. Very stable. Well maintained. Provide the minimal features (split screen, tabs) for a fully working terminal (no need to add another indirection layer, like tmux, that will add complexity to your workflow, or hope that your OS have a window manager that will fill the features gap left by Alacritty).
> 
> I understand the point of Alacritty, their reasons, and it is a great terminal but it seems that Kitty has all of Alacritty's strenghts plus do not lack the minimal set of features to make it a fully working terminal out-of-the-box.

However, kttiy doesn't have persistence across SSH sessions like tmux has. (Not that alacritty has that either).

**Comment by CaninoDev** [comment]
> Kitty is super fast. Very stable. Well maintained. Provide the minimal features (split screen, tabs) for a fully working terminal (no need to add another indirection layer, like tmux, that will add complexity to your workflow, or hope that your OS have a window manager that will fill the features gap left by Alacritty).
> 
> I understand the point of Alacritty, their reasons, and it is a great terminal but it seems that Kitty has all of Alacritty's strenghts plus do not lack the minimal set of features to make it a fully working terminal out-of-the-box.

Where tmux really shines is persistence across sessions (whether remotely or locally). Its ability to manipulate windows and panes is not what draws most people to it.

**Comment by Penguin-Guru** [comment]
Weighing in as a person who was just faced with the decision, I chose Alacritty because:

1. It was much better maintained in my distribution's repository-- Kitty version was not current, had a couple more bugs, had more serious bugs, and was not packaged to be as modular.
2. It didn't depend on as many other packages (most notably Python).
3. Kitty's features do seem great but the emphasis on packing multiple sub-terminals into a single window is redundant since I already use a i3/sway for that.

Sorry if this seems overly critical. I definitely would have chosen Kitty if not for those. Still open to trying it in the future.

**Comment by kenranunderscore** [comment]
I have a couple of points to add that I haven't seen yet:

- I use macOS at work and NixOS or some other GNU/Linux variant at home. Ideally, I'd like to use the "same" terminal emulator on both system, with as few differences in configuration as possible. On paper, both kitty and alacritty would be good candidates. But alacritty has one big flaw (there's a huge issue open for it, but nothing really going on): it does not work correctly when using an alternative (manually created) keyboard layout on macOS; there's something wrong with the underlying library that manages the input, because it seems to react to the correct *scancode*, but not *keycode*. kitty has no such problems on any system that I've tried it on.
- The font rendering seems to be so much better in kitty. Again, on a Linux system this is not as noticeable, but on macOS I have to choose way lighter font variants in alacritty.

I don't even use many of kitty's features, like multiple windows or tabs. Maybe that is why I've yet to run into any problems with it. Great work!

**Comment by junguler** [comment]
i'm stuck on windows as i'm a huge gamer and you guys don't make a windows version, i was so happy when i heard about this project and how it's multi platform and then i saw that it's not on windows and all my hype died instantly.

**Comment by dcow** [comment]
Every time I spin up kitty and Alacritty side-by-side, kitty has prevailed. When kitty launched it was a stable, feature complete terminal emulator. Alacritty had numerous text rendering issues (overlapped characters, incorrect font weights, doesn't respect system DPI scaling settings, flashing lines, you name it). What's the point of fast if the screen doesn't render accurately?

I thought I'd give Alacritty a try again and I run into crap like this: https://github.com/alacritty/alacritty/issues/4673. New versions of Alacritty don't even pass gatekeeper and the maintainers disparage macOS users as idiots who wasted money on Apple hardware (it's in one of the linked issues). Not even a primary macOS user but, nope, sorry, zzz... 

Kitty is a solid, stable, gpu-backed terminal emulator that feels far more polished than Alacritty (although the rendering in Alacritty has improved slightly). Kitty landed solid and has only been maturing over the years. Alacritty still has a long ways to go technically and maturity-wise. Also, kitty's layout support is awesome!

**Comment by gbaranski** [comment]
I'm using alacritty because I don't want features like tabs or panes, because I use either tmux or i3 to do it.

**Comment by horta** [comment]
That is fair, @gbaranski . I guess alacritty and kitty (and iterm2) are just the best terminals in their our niches.

**Comment by trygveaa** [comment]
> I'm using alacritty because I don't want features like tabs or panes, because I use either tmux or i3 to do it.

@gbaranski: How is that a reason? I don't use tabs, but kitty having support for it doesn't affect me in any way.

**Comment by gbaranski** [comment]
> > I'm using alacritty because I don't want features like tabs or panes, because I use either tmux or i3 to do it.
> 
> @gbaranski: How is that a reason? I don't use tabs, but kitty having support for it doesn't affect me in any way.

I don't like having something that I don't use. And I prefer something really minimalistic.

> That is fair, @gbaranski . I guess alacritty and kitty (and iterm2) are just the best terminals in their our niches.

Actually I've used iTerm2 before but switched to alacritty, iTerm2 has a lot of unessesary crap and its much slower. And also it doesn't work on Linux.

**Comment by Sleepful** [comment]
+1 point for Kitty. Alacritty does not have an equivalent option to [this](https://sw.kovidgoyal.net/kitty/conf.html#opt-kitty.macos_option_as_alt):
```
macos_option_as_alt yes
```
[Related issue](https://github.com/alacritty/alacritty/issues/4048)

**Comment by tvidal-net** [comment]
Found this thread and wanted to leave my 2 cents. Found alacritty on DistroTube channel and immediately fell in love with it, but was disappointed that it did not have font-ligatures. Found kitty in another video on the same channel and was also disappointed that it cannot map the right mouse button to paste, and as per #1033, developer seems to treat kitty as a personal project disregarding user feature requests.

Having to choose the best of two evils, I think the right mouse button to paste will be missed a lot more, as middle-clicking to paste often causes undesirable accidental scroll. So I reckon I'll have to live without font-ligatures for now, alacritty is the chosen one.

**Comment by Sleepful** [comment]
Would be interesting to compare how many lines of code each project has.

**Comment by SolitudeSF** [comment]
> Would be interesting to compare how many lines of code each project has.

this is meaningless comparison crosslanguage (and often even in same language).

**Comment by gbaranski** [comment]
> Would be interesting to compare how many lines of code each project has.

Alacritty:
```
--------------------------------------------------------------------------------
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 Rust                    66        24253         3740         2642        17871
 Markdown                 7         1704          399            0         1305
 Plain Text               1          779           58            0          721
 Bourne Shell             4          196           16           24          156
 Toml                     5          155           20            0          135
 JSON                   120          120            0            0          120
 Makefile                 1           66           12            0           54
 XML                      1           36            0            3           33
 YAML                     1          812           71          741            0
--------------------------------------------------------------------------------
 Total                  206        28121         4316         3410        20395
--------------------------------------------------------------------------------
```
Kitty:
```
--------------------------------------------------------------------------------
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 C                       72        44551         4837         2227        37487
 Python                 173        38011         5952         3030        29029
 C/C++ Header            56        27430         1381         7575        18474
 Objective-C              7         5995          845          377         4773
 reStructuredText        32         6861         2099            0         4762
 Plain Text               1         3692            0            0         3692
 JSON                     3          547           32            0          515
 HTML                     5          110           14            1           95
 CSS                      1          126           28            7           91
 Makefile                 2           63           16            8           39
 Markdown                 2           28            9            0           19
 VimL                     1           17            0            1           16
 Bourne Shell             1           30            3           11           16
--------------------------------------------------------------------------------
 Total                  356       127461        15216        13237        99008
--------------------------------------------------------------------------------
```
I thought that Kitty is written mostly in Python

**Comment by SolitudeSF** [comment]
@gbaranski you are also counting bundled glfw

**Comment by gbaranski** [comment]
Excluding `glfw/*` dir
```
--------------------------------------------------------------------------------
 Language             Files        Lines        Blank      Comment         Code
--------------------------------------------------------------------------------
 Python                 171        37710         5904         2961        28845
 C                       44        26833         2357          783        23693
 C/C++ Header            32        17276          716         2015        14545
 reStructuredText        32         6861         2099            0         4762
 Plain Text               1         3692            0            0         3692
 Objective-C              2         1288          162           33         1093
 JSON                     2          395           32            0          363
 HTML                     5          110           14            1           95
 CSS                      1          126           28            7           91
 Makefile                 2           63           16            8           39
 Markdown                 2           28            9            0           19
 Bourne Shell             1           30            3           11           16
 VimL                     1           17            0            1           16
--------------------------------------------------------------------------------
 Total                  296        94429        11340         5820        77269
--------------------------------------------------------------------------------
```

And also, the reason why I would prefer Alacritty than Kitty, is because from my experience most of the thing that are written in Python don't work out of the box and require some workaround.

**Comment by Sleepful** [comment]
> > Would be interesting to compare how many lines of code each project has.
> 
> this is meaningless comparison crosslanguage (and often even in same language).

It is not so meaningless if you contextualize the results, weighing in differences in features, comparing the parts that do the same thing, etc. 
If a project is three times as large than another project equally capable then that means that you need to read three times the amount of code to fully grok it. Obviously things like readability are important, at the end of the day it is one data point that tells me to be wary of getting too involved if I'm not sure I can navigate 70k loc, with a good share of C code.

**Comment by dcow** [comment]
> Found this thread and wanted to leave my 2 cents. Found alacritty on DistroTube channel and immediately fell in love with it, but was disappointed that it did not have font-ligatures. Found kitty in another video on the same channel and was also disappointed that it cannot map the right mouse button to paste, and as per #1033, developer seems to treat kitty as a personal project disregarding user feature requests.
> 
> 
> 
> Having to choose the best of two evils, I think the right mouse button to paste will be missed a lot more, as middle-clicking to paste often causes undesirable accidental scroll. So I reckon I'll have to live without font-ligatures for now, alacritty is the chosen one.

- [ ] Both Kitty and Alacritty have abrasive maintainers that promptly close issues and suggestions using a super dismissive tone that makes it hard for community members to contribute in a meaningful and positive way. But at least Kitty's maintainer doesn't actively shit on macOS users. On the flip-side Kitty burnt a bridge with the term-db folks because apparently "everything in the world breaks if it's not xterm" so you still have to copy the terminfo to every new host you use. Gets old fast. I guess it's a lesser of two evils thing for me, too.

Also a side note: both Alacritty and Kitty have a solo maintainer who isn't staffing the project full time. IMHO they could really benefit from the resources available in their communities if they'd only be more receptive to collaboration. It is what it is and sadly seems to be a trend with terminal emulator authors...

**Comment by kovidgoyal** [comment]
Large parts of the source code in kitty is auto-generated. If you really want to count lines of code use the included ./count-lines-of-code script, though what lines of code could possibly tell you, I have no idea. An extremely small project could be hard to contribute to and an extremely large project easy, depending on code organization/structure/contribution policies, etc. According to @dcow he cant contribute because I am abrasive, not because of lines of code :roll_eyes: 

```
     215 text files.
     215 unique files.                                          
      33 files ignored.

github.com/AlDanial/cloc v 1.88  T=0.29 s (701.3 files/s, 201342.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                         122           5057           2372          21806
C                               39           2143            749          20055
C/C++ Header                    26            453            705           2763
Objective-C                      2            159             33           1083
GLSL                            12             91             70            451
AsciiDoc                         1              0              0              1
-------------------------------------------------------------------------------
SUM:                           202           7903           3929          46159
-------------------------------------------------------------------------------
```

**Comment by programus** [comment]
I think the reason may be the bad support of tmux. 
https://github.com/kovidgoyal/kitty/issues/249
https://github.com/kovidgoyal/kitty/issues/1118
https://github.com/kovidgoyal/kitty/issues/377

In my case, I decided to move to some other better terminal emulator from iTerm2 which is becoming slower and slower recently. 
I tried alacritty since it says itself fast. 
Since it has no features of tabs, split panes, I decided to use tmux in alacritty. I am not so familiar with tmux, but after some learning, I love it now. And I am now using oh my tmux, which have many preset settings I love. 
Also since it has no quake-style shortcut, I created [this](https://gist.github.com/programus/d3bf9c6b8f9b7648891d1f59b665b35e) based on hammerspoon to make it work just like iTerm2. 
(I think this can also be used on kitty with some tiny modification. )

After that, I found alacritty does not support font ligature, which is not critical for me but I wish there was something better. So I found kitty which should be also fast I think and give it a try.
I tried kitty but it disappoint me once I started tmux. The tmux does not work at all if I enable mouse. Not only the wheel and scrolling, the cursor is always staying at the left bottom corner and the statusbar sometimes does not appear at all. 
Obviously, this is a critical problem for me. 
So I searched on Internet and got the issues above. I found none of these can solve my problem and the author of kitty does not have willing to solve these problems at all and keep claim these are bugs of tmux. 
It may be a bug of tmux, but since the problems happen only on kitty, I hope kitty can give others a hand and communicate with tmux community to find out what the reason is instead of just close the issues. 

So I have to choose alacritty, at least for this time, because I cannot see any hope that my problem would be solved. After all, lacking of font ligature is just a nice-to-have feature for me but tmux has almost become a necessary part for me. 
I would have abandoned iTerm2 earlier if I knew tmux is so powerful and not that hard to use. 

Since tmux is a great software with a huge group of users, the compatibility problem might make many users like me to select alacritty rather than kitty. 

I hope one day, kitty can improve, and I might come back. That might be the day kitty is better known than alacritty. 
Anyway, I really like the easy to remember name of kitty much much more than the extremely hard to remember name of alacritty. That's another reason I assigned a shortcut for it instead of start it from spotlight, because I cannot remember the name.

**Comment by junguler** [comment]
my only issue with kitty was that it's not on windows, i have switched to linux recently and there is no other reason not to use kitty as it has way more features than alacritty. however i have found a good alternative to both of these in wezterm, it has all the features that i want (mainly tab and pane support that alacritty lacks) and it's easier to configure and get up and running. i'm sure i'll come back to kitty at some point but for the time being wezterm is my sweet spot between kitty and alacritty.

**Comment by g6ai** [comment]
@programus I have been using tmux in kitty without a problem, including mouse and scrolling. My config is here: https://github.com/g6ai/dotfiles

I've read the issues you mentioned, I noticed people are not really providing their tmux config so no way to locate the problem. If you can share your config, maybe we can figure out a solution.

**Comment by yashank09** [comment]
I've been using Alacritty for more than a year now, and it gives me exactly what I want so I stick to it. With a tabbed WM/tmux, and by using `alacritty-ligatures` from the AUR repo, I worked out most of the problems people have with Alacritty. (alacritty-ligatures on the AUR is sick btw! works flawlessly so shoutout to the developer(s))

However I recently got rid of i3, and my current WM does not support tabs. I thought this wouldn't be a problem but oh boy, aren't terminal tabs are a savior. This requirement and seeing 'Alacritty vs Kitty' discussions on every other git issue, my first goto choice had to be Kitty. 

Today would be my 3rd day of using Kitty, and holy hell it is fast as shit... Not just fast, but the boot time and execution speed are noticeably faster than Alacritty. Not sure if this other thing is a placebo, but apparently the fonts seem to be rendering much clearer on Kitty too?? 
As for the switch, I was able to transfer a year's worth of customization onto Kitty in a matter of minutes. My whole experience has been fantastic so far! 

Looking a year back, the only reason I chose Alacritty over Kitty was because of Rust and my bias towards its fad. This, along with how Alacritty is marketed and presented, would be the only reason one would choose Alacritty. 
I thought I had settled on a perfect terminal, but sometimes, perfection is just a tabless WM away :) 

Thanks @kovidgoyal !

**Comment by g6ai** [comment]
@yashank09 My journey is similar to you. I had been using Alacritty for a while, for me the deal breaker is powerline font rendering, curly/coloured underline (also ligature, I had no idea there's `alacritty-ligatures` on AUR so thanks for the info). kitty does a good thing to my eyes. Right now the only thing I miss about Alacritty is its brilliant icon :)

**Comment by kovidgoyal** [comment]
@g6ai Not sure on what basis you have chosen your list of features to test, but here are a few more, which most serious terminal emulators implement, and alacritty does not. hyperlink support, scripting (which in kitty is possible via both kittens and remote control), tiling/tabbing, opening scrollback buffers in external programs, startup sessions and session save/restore (the latter of which kitty does not yet support), live search, unicode input, input method extensions (which I think alacritty supports some of), image display, a client/server model (which in kitty is --single-instance), etc.

**Comment by g6ai** [comment]
@kovidgoyal It's quite arbitrary from my own experience as a user. Thank you for the details, I'll look into it as I learn!

**Comment by luisdavim** [comment]
Alacritty does support hyperlinks...

**Comment by kovidgoyal** [comment]
No, it does not: https://github.com/alacritty/alacritty/issues/922

**Comment by luisdavim** [comment]
I use Alacritty and I can `cmd+click` links displayed in the terminal to open them in a browser:

```yaml
mouse:
  # Click settings
  #
  # The `double_click` and `triple_click` settings control the time
  # alacritty should wait for accepting multiple clicks as one double
  # or triple click.
  #double_click: { threshold: 300 }
  #triple_click: { threshold: 300 }

  # If this is `true`, the cursor is temporarily hidden when typing.
  hide_when_typing: true

  url:
    # URL launcher
    #
    # This program is executed when clicking on a text which is recognized as a URL.
    # The URL is always added to the command as the last parameter.
    #
    # When set to `launcher: None`, URL launching will be disabled completely.
    #
    # Default:
    #   - (macOS) open
    #   - (Linux/BSD) xdg-open
    #   - (Windows) explorer
    #launcher:
    #  program: xdg-open
    #  args: []

    # URL modifiers
    #
    # These are the modifiers that need to be held down for opening URLs when clicking
    # on them. The available modifiers are documented in the key binding section.
    modifiers: Command
```

what more support do you need?

**Comment by kovidgoyal** [comment]
Go read about terminal hyperlinks. I can run grep to search for
something and then click on the results to open them in my editor at the
correct line.
https://sw.kovidgoyal.net/kitty/kittens/hyperlinked_grep.html

And here some more things you can do in kitty and most serious
terminals, and not in alacritty.

Under the category: "Crippled without tmux"
    1) Use multiple windows/tabs without wasting CPU cycles/energy passing every byte
    through a middleman
    2) Use multiple windows with native scrolling
    3) Use multiple windows with proper mouse selection support

Under the category: "Incomplete font rendering"
    1) No ligatures
    2) Font glyphs bleed into neighboring cells when they dont fit
    3) No unicode standard based width implementation, which means th
    width a glyph takes has no standards based representation 

Under the category: "General lack of features"
    1) No image display
    2) No way to pipe screen and scrollback into external programs
    3) No way to script the terminal, at all. 
    4) No client/server, so no instant startup, and wasted memory usage
    5) No way to define startup sessions
    6) No way to interact with the screen using only the keyboard, for
    example, you cant click links using the keyboard

Under the category: "Performance"
    1) Poor latency
    2) Overblown claims of throughput 
    3) No way to tune performance tradeoffs to suit individual
    preferences

And hey if you want to use Alacritty, suit yourself. Just take a little
time to educate yourself on what you are missing. Or not, it's your life.

**Comment by luisdavim** [comment]
For me the main issue with Kitty is the lack of tmux support. Most of the rest are nice to have things that I don’t really miss. 
Most serious terminal emulators do support tmux...

**Comment by ctrlcctrlv** [comment]
I use tmux in Kitty all the time. In what way is it not supported?

[Certain caveats](https://sw.kovidgoyal.net/kitty/faq.html#i-am-using-tmux-and-have-a-problem) exist, which you should keep in mind, but they are well documented.

**Comment by kovidgoyal** [comment]
Indeed anything you can do in alacritty with tmux you can do in kitty with tmux. The problems with tmux come from the tmux maintainer refusing to add support for the various advanced things kitty does.

**Comment by kovidgoyal** [comment]
And  again, unlike alacritty one of kitty's goals is to move the entire terminal ecosystem forward by finally ending the decades to stagnation in this space. See https://sw.kovidgoyal.net/kitty/protocol-extensions.html

**Comment by horta** [comment]
Is there no committee discussing those protocol standards? https://unix.stackexchange.com/questions/5800/what-protocol-standard-is-used-by-terminals

Having a common minimum set of features that are revised over the years would useful to everyone developing a new terminal.

**Comment by kovidgoyal** [comment]
Nope, there's herding cats and then there's herding terminal developers. I prefer cats :))  The last attempt I took part in was https://gitlab.freedesktop.org/terminal-wg now totally dead.

**Comment by sujay1844** [comment]
Reasons I prefer Kitty over alacritty:
- **Kitty is made by an Indian!!**

**Comment by programus** [comment]
> @programus I have been using tmux in kitty without a problem, including mouse and scrolling. My config is here: https://github.com/g6ai/dotfiles
> 
> I've read the issues you mentioned, I noticed people are not really providing their tmux config so no way to locate the problem. If you can share your config, maybe we can figure out a solution.

Thank you for your hint. Finally, I found out the reason. It is because I used a unicode character,🖱 , to indicate the mouse status in the status bar. I changed the character, then everything work just fine. 

Now, I think I will give kitty a try. 

I made a [hammerspoon script](https://gist.github.com/programus/d3bf9c6b8f9b7648891d1f59b665b35e) for alacritty to enable it quake style console, I just want to know whether kitty have this settings or I have to create another one for kitty.

**Comment by Luflosi** [comment]
kitty does not have such a setting, see #45.

**Comment by programus** [comment]
> kitty does not have such a setting, see #45.

Thank you. I also realize this and have worked out [my version](https://gist.github.com/programus/f8b6199876e693a27969f262469284ad), which supports multiple monitors. 

So far, kitty is really great!

**Comment by infokiller** [comment]
> > Found this thread and wanted to leave my 2 cents. Found alacritty on DistroTube channel and immediately fell in love with it, but was disappointed that it did not have font-ligatures. Found kitty in another video on the same channel and was also disappointed that it cannot map the right mouse button to paste, and as per #1033, developer seems to treat kitty as a personal project disregarding user feature requests.
> > Having to choose the best of two evils, I think the right mouse button to paste will be missed a lot more, as middle-clicking to paste often causes undesirable accidental scroll. So I reckon I'll have to live without font-ligatures for now, alacritty is the chosen one.
> 
> Both Kitty and Alacritty have abrasive maintainers that promptly close issues and suggestions using a super dismissive tone that makes it hard for community members to contribute in a meaningful and positive way.

IMO the "super dismissive tone" is your personal interpretation. I agree the communication could be improved, but as someone who opened 3 issues in Kitty (and read others), I can say @kovidgoyal not only responds very fast to issues, but is also making a big effort to help people that open issues, and is very open to PRs and external contributions.

**Comment by tasuki** [comment]
Just a point of view...

@Penguin-Guru chose Alacritty because it was better maintained in his distribution's repository - I chose Kitty for a similar reason: it _is_ in my distribution's repository.

@kovidgoyal yes you can be a little abrasive, but you're the best Kovid we have, and we still love you and are very grateful for all your work :heart:

**Comment by AtaraxianWolf** [comment]
I'm a Debian guy but I started a new position a couple of months ago where they gave me a MacBook Pro, and I liked it so much I went out and bought myself an M1 MBA. Being someone who had hardly used OSX before this I went through a lot of terminals before I settled on kitty, and use it as my daily driver both at work and at home.

**Comment by horta** [comment]
I was going through kitty source code and I see that @kovidgoyal makes use of python C API a lot. Going back to the first commits it seems it was mostly Python code and then overtime became a mix of C and Python.

Do you plan to migrate to C only (with no python dependency), @kovidgoyal ? Or is there a non-historical reason for keeping the mix?

**Comment by kovidgoyal** [comment]
Python is perfect for the UI and scripting parts of kitty. So its always
going to be there. In fact it's very common that non-trivial
applications written in compiled languages eventually get a scripting
component bolted on. With kitty it's builtin.

The kittens, layout system, remote control, etc are all powered by
python, which is the right tool for those jobs. 

The only real downside to python is the 50-100ms startup time cost it
imposes.

**Comment by ctrlcctrlv** [comment]
I wonder if `kitty` could be launched with some parts of it waiting on Python to be initialized to run? That'd work for most users, since most users don't immediately launch a kitten, but just launch a shell.

If their config file _does_ call some Python part before it's initialized, just make the thread wait until Python is initialized. Same as before.

If you think this is worth pursuing I'm happy to think about it more/even PR. Kitty has been back to being my daily driver for the last 6 months, so I've been meaning to get back to contributing. (Been busy with my [from scratch font editor project](https://github.com/MFEK), which Kitty is used to develop!)

**Comment by horta** [comment]
Thanks for the answer, @kovidgoyal !

Do you know about the Lua language? It seems quite a good fit for such cases. (In case you ever consider another scripting language.)

**Comment by kovidgoyal** [comment]
On Wed, May 12, 2021 at 05:17:58AM -0700, Danilo Horta wrote:
> Thanks for the answer, @kovidgoyal !
> 
> Do you know about the Lua language? It seems quite a good fit for such cases. (In case you ever consider another scripting language.)

Yes I know it, it's nice enough, however when designing a scripting
language for an application to be used by users I think popularity 
and ease of use are the most important metrics so that it can be used by
the most people the most easily, and Python or JS would be better fits
there, except that I personally cant stand JS, I even wrote a transpiler
of my own to make apython like language that compiles to JS
https://github.com/kovidgoyal/rapydscript-ng

**Comment by kovidgoyal** [comment]
On Wed, May 12, 2021 at 04:46:50AM -0700, Fredrick Brennan wrote:
> I wonder if `kitty` could be launched with some parts of it waiting on Python to be initialized to run? That'd work for most users, since most users don't immediately launch a kitten, but just launch a shell.

Personally, I dont see the need, the startup delay imposed by python is
almost negligible on human scale and if not --single-instance makes it
truly instant. But hey if that's something you like to pursue, feel free
to investigate. You will have a tough time since kitty has python pretty
deeply interwoven into ts guts as it powers the UI.

**Comment by shiny-comic** [comment]
Why I don't use alacritty is still no support for inline japanese input and people be like "just use english".

**Comment by Kasahs** [comment]
I have been using Alacritty on my M1 Macbook Pro for a few months now. Dealing with issues like [this](https://github.com/alacritty/alacritty/issues/4410). Which according to the dev has something to do with MacOS BigSur's OpenGL implementation.
Have been waiting on a fix for a while but the issue and others like it have been closed. No resolution.

No such problem with Kitty. Plus it starts faster, runs faster, has better font rendering and I don't know if it's just me but better color accuracy. I don't care for the pane management features, but with tmux kitty runs just fine for me and way feels much better than Alacritty.

Looks like Rust won't be magically saving us after all.

**Comment by kovidgoyal** [comment]
On Fri, May 21, 2021 at 02:25:03AM -0700, Shasak Raina wrote:
> I have been using Alacritty on my M1 Macbook Pro for a few months now. Dealing with issues like [this](https://github.com/alacritty/alacritty/issues/4410). Which according to the dev has something to do with MacOS BigSur's OpenGL implementation.
> Have been waiting on a fix for a while but the issue and others like it have been closed. No resolution.

Wow, I feel for the alacritty devs, if someone reported a similar issue
with kitty, I too wouldn't have a clue what could cause it, or how to
remedy it. If I had to guess I would say CPU <-> GPU bandwidth is the
culprit.

That said, welcome to the dark side :)

**Comment by darthrevan13** [comment]
I'm amazed there is still no MS Windows support for kitty. Allacritty has some bugs there but still, at least it's usable and one can have the same terminal emulator and mostly the same config file across all OSes.

**Comment by ctrlcctrlv** [comment]
@darthrevan13 Thank you for the reminder. This has been on my todo. I was waiting for WSL2 to come out, see #640. It seems that it has, see [microsoft/wslg](https://github.com/microsoft/wslg), which is very promising as it includes GPU drivers and this promising diagram:

![image](https://user-images.githubusercontent.com/838783/120441920-152e7480-c353-11eb-82b2-c461eeac7a2f.png)

As I stated in #640, the main problem with Windows support was the fact that somebody would need to do a deep dive into Kitty's GLFW fork and make it Windows-compatible, and then maintain that into perpetuity.

I'd encourage you to try running Kitty under WSL2 and report your findings in #640. If the image is correct, it's already supported. Just needs someone like you to document the build. I don't know where the charger is for my Windows laptop right now, so probably won't be me. Plus this is my last email of the night, lucky you :-)

**Comment by ctrlcctrlv** [comment]
I'm a bit of an insomniac it seems. 

It's my understanding that PowerShell can be compiled for Linux. Am I missing the point, or does that meet your needs? Someone was telling me to try it. 

Anyway, I've used Windows before and never found a way to use one shell no matter what. I always needed to open cmd occasionally, couldn't just always use mintty, which is what I currently would use on Windows, or its tabbed cousin fatty.exe

**Comment by Neurognostic** [comment]
[Debate](https://github.com/alacritty/alacritty/issues/910#issuecomment-855464558) on Alacrity project about which graphics protocol should be supported.
 
> [kitty's protocol has a lot of undefined behavior and Linux-isms (e.g. shared memory) that render it IMHO a dead end. ](https://github.com/alacritty/alacritty/issues/910#issuecomment-855464558)

Any idea what "undefined behavior" the Kitty graphics protocol has that makes it unsuitable to be implemented in Alacritty? Someone with more knowledge about Kitty's Graphics Protocol ought to chime in there, as we should probably be working towards a unified protocol (or at least interface) across all terminal-emulators.

**Comment by kovidgoyal** [comment]
On Tue, Jun 08, 2021 at 04:53:17PM -0700, Donovan Tremura wrote:
> Any idea what "undefined behavior" the Kitty graphics protocol has that makes it unsuitable to be implemented in Alacritty? Someone with more knowledge about Kitty's Graphics Protocol ought to chime in there, as we should probably be working towards a unified protocol (or at least interface) across all terminal-emulators.

I have responded to those nonsense allegations.

**Comment by mibli** [comment]
This discussion is why I stick with rxvt-unicode, thanks guys!

**Comment by ctrlcctrlv** [comment]
:star::star::star::star: 

What's that?

Total number of stars your projects have received.

Spend more time trying to increase them and less time trolling Kitty's issue tracker.

Then, maybe, someone will care which tools you use and don't use. Don't be a thought leader with no followers, it's embarassing.

**Comment by mibli** [comment]
> What's that?
> 
> Total number of stars your projects have received.
> 
> Spend more time trying to increase them and less time trolling Kitty's issue tracker.
> 
> Then, maybe, someone will care which tools you use and don't use. Don't be a thought leader with no followers, it's embarassing.

Do You really want to get personal, especially on star-whoring? ;) I'm not here for that. I'm just noticing, that threads like these, which point out serious issues with both of the applications, help to deal with mature projects like urxvt, in which the stuff just works, and it works fast, even if it is dated. And I just honestly wanted to thank You all, for pointing them out and preventing me from wasting precious hours on trying out the new trendy things :)

**Comment by SolitudeSF** [comment]
there is no terminal in which everything "just works". not even/especially urxvt.

**Comment by yashank09** [comment]
At this point I feel this thread should be closed or only opened to discussion by contributors.

**Comment by clayrisser** [comment]
> Alacritty runs on Windows and is the default terminal emulator on Sway.

Actually foot is the default terminal on sway.

https://codeberg.org/dnkl/foot

**Comment by Luflosi** [comment]
The default was changed recently: https://github.com/swaywm/sway/commit/009c58fc9529e284569bf58b73964cb3dc7c5baf.

**Comment by clayrisser** [comment]
@Luflosi I don't like the foot terminal personally. It has lots of bugs and does not display characters correctly. Plus the image support of kitty is sick.

**Comment by shezaan** [comment]
I think the advantage for Alacritty from a pure language standpoint is that the use of Rust makes it more secure... unless they are using "unsafe" in a lot of places. I haven't looked at their code but speaking from a theoretical PoV.
Kitty definitely seems faster, but it seems too featureful vis-a-vis more prone to bugs, issues etc. There is also a lot of languages used python, C, objective-C etc, which theoretically should make it more bloated and thereby slower but somehow doesn't? Alacritty has Rust as its main language so its should be memory safe.
With Kitty, I can't seem to get the basics right, and that really must be my fault, but "bold" fonts don't look the same on Kitty vs every other terminal emulator I've used. On Kitty, bold doesn't look "bold" at all. I've only just started using Kitty so I'm still discovering it. But for now Alacritty and Kitty seem pretty solid to me as I haven't run into issues with either one of them.

**Comment by kovidgoyal** [comment]
On Mon, Jul 19, 2021 at 07:11:53AM -0700, shezaan wrote:
> I think the advantage for Alacritty from a pure language standpoint is that the use of Rust makes it more secure... unless they are using "unsafe" in a lot of places. I haven't looked at their code but speaking from a theoretical PoV.

With experience one learns that language choices are 
the least relevant thing to the success/failure/quality 
of non-trivial software. 

> Kitty definitely seems faster, but it seems too featureful vis-a-vis more prone to bugs, issues etc. There is also a lot of languages used python, C, objective-C etc, which theoretically should make it more bloated and thereby slower but somehow doesn't? Alacritty has Rust as its main language so its should be memory safe.

Code size and feature counts are not correlated to speed or bugs.
What matters for speed, is the actual code in the hot path, to which how
much other code exists is largely irrelevant. What matters for bugs is
bug *density* not counts, with more features, bug counts will obviously
be higher, but bug density may or may not be, depending on code quality.

> With Kitty, I can't seem to get the basics right, and that really must be my fault, but "bold" fonts don't look the same on Kitty vs every other terminal emulator I've used. On Kitty, bold doesn't look "bold" at all. I've only just started using Kitty so I'm still discovering it. But for now Alacritty and Kitty seem pretty solid to me as I haven't run into issues with either one of them.

Use --debug-font-fallback to check what font is actually being used for
bold.

**Comment by dcow** [comment]
> I think the advantage for Alacritty from a pure language standpoint is that the use of Rust makes it more secure... unless they are using "unsafe" in a lot of places. I haven't looked at their code but speaking from a theoretical PoV.
> 
> Kitty definitely seems faster, but it seems too featureful vis-a-vis more prone to bugs, issues etc. There is also a lot of languages used python, C, objective-C etc, which theoretically should make it more bloated and thereby slower but somehow doesn't? Alacritty has Rust as its main language so its should be memory safe.

You betray your lack of understanding of how a sophisticated software project works. Almost everything ends up being a polyglot as it matures. You generally want to use the right tool for the job rather than shoehorn everything into a single language.

**Comment by virus-found** [comment]
I chose alacritty over kitty, because the former was more kiss.
Now moved to foot terminal, because I sensed the development of alacritty to become less and less kiss.

**Comment by c02y** [comment]
When I chose standalone terminals, I chose alacritty and kitty, since I love Rust, I chose alacritty as my daily terminal at first, had been using it for a few months, really disappointed, you can search "sluggish" in alacritty issue page you'll find over ten issues. Personally, no such issue in kitty at all.

I was annoyed by the sluggish issue of alacritty over and over again, and I found a long command `xdotool key --delay 0 $(for i in $(seq 1 10); do for j in $(seq 0 9); do echo -n "$j "; done; done)` from alacritty issue page, I tried it in alacritty and kitty, it freezes my alacritty(empty config) completely. Kitty works just fine.

Finally, I replaced alacritty with kitty as my daily terminal, I am pretty happy about it so far.

Since kitty is written in Python and alacritty is written in Rust, I thought at the beginning that alacritty should have better performance than kitty, it turns out I was wrong, I tested the performance of basic commands on both of them when I started to choose standalone terminals, kitty and alacritty have similar performance, and the two issues on alacritty I just mentioned suggest that actually, kitty has better performance, it really surprises me. 

Since kitty is written in Python, the author @kovidgoyal and co-developers really are talented in Python and did a lot of work on the performance. Thank you.

**Comment by SolitudeSF** [comment]
> kitty is written in Python

performance sensitive parts are written in `C`

**Comment by shaunsingh** [comment]
As a macOS user, I've been through quite a few terminals and stuck with kitty in the end. The Pros and Cons of each I've found: 

(note that tabs or splits aren't an issue, since I use tmux)

Iterm2: Nice terminal, looks good, but 
1. had issues rendering characters in neovim
2. can't be configured through the terminal, making it a pain with dot files 
3. feels very slow 

Alacritty: Felt much much faster, looked good as well, but 
1. Doesn't have ligatures at all 
2. Doesn't have a proper implementation of   windows (opening another window just another instance of the app, and clutters everything up 
3. No image support 

Wezterm: Didn't launch 
Terminal.app: Felt the fastest, but no 24-bit color

Kitty is the only terminal out there that 1. has ligature support, 2. feels fast, 3. has image support, and 4. can be configured through dot files. 

However, I'm still on the search for a better terminal. Compared to Iterm or Alacritty, font rendering looks worse, transparency doesn't work with a transparent menubar (whereas in iterm you can make everything transparent), scrolling is less responsive compared to Alacritty. I feel like Alacritty could evolve into my perfect terminal emulator, but the developer doesn't like macOS very much, and even refuses to just change the icon. 

These days there is no perfect terminal emulator for macOS, its just about how many compromises you're willing to accept. I wish there was an equivalent to the ST Terminal, where you start with an extremely basic terminal you can patch up to your needs.

**Comment by kovidgoyal** [comment]
Use macos_thicken_font if you want thicker fonts. And I have no idea what scrolling is less responsive means. If you mean scrolling the scrollback buffer, that is instant, in pretty much every terminal out there. If you mean scrolling in terminal programs, I cannot reproduce, and you might want to adjust input_delay in kitty.conf.

**Comment by puhitaku** [comment]
I was looking for where I should post my love to Kitty. I began to patronage a bit of my money for the author today. Keep it up, and I'll spread the name ❤️

**Comment by kovidgoyal** [comment]
Thanks :)

**Comment by dlipovetsky** [comment]
For what it's worth, I have used the https://pavelfatin.com/typometer/ tool to measure the _visual latency_ of alacritty, kitty, gnome-terminal on my machine. As far as I can tell, both kitty and alacritty are utilizing my GPU. I'll let my results speak for themselves:

![image](https://user-images.githubusercontent.com/3445370/131775143-458efcd8-04b7-42eb-ac5b-ac0c578bc47a.png)

**Comment by kovidgoyal** [comment]
Wow, given that thanks to, https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.input_delay kitty deliberately introduces 3ms of delay when processing input, that mean a 5ms time from keypress to pixels on screen. I am pleased it is possible to achieve these kinds of speeds on a modern machine with all its layers of indirection.

**Comment by dlipovetsky** [comment]
I did not know about `input_delay`! :smile:  Repeating my test with`input_delay 0`:

![image](https://user-images.githubusercontent.com/3445370/132046140-7660684c-1c28-44ad-ac65-33aec2c48240.png)

**Comment by duarm** [comment]
I do everything on my terminal, and open hundreds of terminals per day on a tiling window manager, for that reason I prefer minimal terminals without the gpu context creation overhead. But if I would ever switch to a gpu terminal, I would favor startup over most things. Alacritty has about 25% faster startup in comparison with kitty, I bet it's because of python's initialization, not sure tho. Just bringing that up since no one discussed startup time so far.

**Comment by horta** [comment]
It is python indeed. Python has a slow startup specially if you actually import packages (which is of course necessary).

But once Kitty process is running, there is no noticeable delay to open new windows and tabs.

**Comment by shaunsingh** [comment]
> Use macos_thicken_font if you want thicker fonts.

I am using it, font rendering still looks worse. Thicker isn't what I'm looking for, alacritty looks sharper 

> And I have no idea what scrolling is less responsive means. If you mean scrolling the scrollback buffer, that is instant, in pretty much every terminal out there. If you mean scrolling in terminal programs, I cannot reproduce, and you might want to adjust input_delay in kitty.conf.

Responsive as in there isn't a big jutter when I start scrolling, I'll mess around with input_delay though, thanks

**Comment by trygveaa** [comment]
> I do everything on my terminal, and open hundreds of terminals per day on a tiling window manager, for that reason I prefer minimal terminals without the gpu context creation overhead. But if I would ever switch to a gpu terminal, I would favor startup over most things. Alacritty startup time is faster (~0.3s/term) than kitty (~0.4s/term), (minimal terminal just for comparison is ~0.044s/term, I'm not switchin that soon), I bet it's because of python's initialization, not sure tho. Just bringing that up since no one discussed startup time so far.

If you want kitty to open fast you should use the `--single-instance`/`-1` option. With that all subsequent windows (when kitty is already running) open faster than alacritty.

**Comment by horta** [comment]
@kovidgoyal seems adamant to stick with python for certain tasks. I had a look at the python bits and it makes use of Python C API, which I find daunting. I wonder if translating Python parts to C (and therefore avoiding Python C API) would allow more people to contribute to the project.

**Comment by duarm** [comment]
> > I do everything on my terminal, and open hundreds of terminals per day on a tiling window manager, for that reason I prefer minimal terminals without the gpu context creation overhead. But if I would ever switch to a gpu terminal, I would favor startup over most things. Alacritty startup time is faster (~0.3s/term) than kitty (~0.4s/term), (minimal terminal just for comparison is ~0.044s/term, I'm not switchin that soon), I bet it's because of python's initialization, not sure tho. Just bringing that up since no one discussed startup time so far.
> 
> If you want kitty to open fast you should use the `--single-instance`/`-1` option. With that all subsequent windows (when kitty is already running) open faster than alacritty.

That's nice, I wonder how kitty would compare with alacritty when both terminals have that feature, apparently they're still working on that feature on alacritty.

**Comment by oblitum** [comment]
I've been using Alacritty for a very long time, before learning of Kitty's existence. I wish I could try Kitty due to its implementation of colored and styled undercurls, [which Alacritty doesn't support](https://github.com/alacritty/alacritty/issues/1628), at the moment that's the sole thing that attracts me to Kitty. I don't give it a go though because that doesn't outweigh one killer feature of Alacritty for me, which is: extensive fontconfig support. Some people despise font configurations through fontconfig, but I learned to appreciate it notice how flexible it can be, and it's that flexibility which I eventually became attached to and have no idea how to replicate in Kitty's conf (if someone knows how, please share given I couldn't infer how from Kitty manual pages):

`~/.config/fontconfig/fonts.conf`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE fontconfig SYSTEM "/usr/share/xml/fontconfig/fonts.dtd">
<fontconfig>
    <alias>
        <family>Terminus</family>
        <prefer>
            <family>TerminessTTF Nerd Font</family>
            <family>Noto Color Emoji</family>
        </prefer>
    </alias>
    <match target="font">
        <test name="family"><string>Terminus</string></test>
        <edit name="antialias" mode="assign"><bool>false</bool></edit>
    </match>
    <match target="pattern">
        <test name="family"><string>Terminus</string></test>
        <test name="slant" compare="more"><const>roman</const></test>
        <edit name="family" mode="assign" binding="strong"><string>IBM Plex Mono</string></edit>
    </match>
</fontconfig>
```

So, let me explain what that does. On ArchLinux I like the [Terminus font](https://archlinux.org/packages/community/any/terminus-font/), but this font is  very bare without much Unicode or emoji glyphs at all. I like to use this font without any antialias too, to make for a very crispy and sharp bitmap font (specially important in ≤1080p, non-retina displays and such). I also do wish to have plenty support for Unicode and Nerd Font chars, so I pick the Terminus Nerd Font variant ([TerminessTTF](https://aur.archlinux.org/packages/nerd-fonts-terminus/)) as a fallback. For such Unicode and Nerd Font glyphs though, I do wish to keep antialias (to avoid getting jagged nerd glyphs and such). At the end of the chain, I pick Noto Color Emoji as emoji fallback. I also pick IBM Plex Mono for italics.

**Comment by kovidgoyal** [comment]
kitty uses fontconfig too. Your problem is using a bitmap font. bitmap fonts are not supported in
kitty.

**Comment by ctrlcctrlv** [comment]
@oblitum "bitmap fonts are not supported in kitty." has a gigantic asterisk…
![image](https://user-images.githubusercontent.com/838783/140284006-5a1d1f51-58fa-4ba1-a2ea-721217b37f9b.png)
![image](https://user-images.githubusercontent.com/838783/140284468-36233518-d62f-4b3b-a860-7538b765ec9e.png)
![image](https://user-images.githubusercontent.com/838783/140284533-17324fcf-2523-4578-aef1-35c6bfb9720a.png)

See https://github.com/ctrlcctrlv/bitmapfont2otb and https://github.com/kovidgoyal/kitty/issues/97#issuecomment-579094656

**Comment by shezaan** [comment]
I'm a bit late to this discussion but these are my thoughts:
Alacritty:
1. I prefer how alacritty renders the fonts, they look bolder and less spaced out. I don't know much about fonts but I think Alacritty's rendering just looks a lot better. =
2. Alacritty works on Windows
3. Alacritty, being written in Rust might likely be more memory-safe than Kitty, this is what initially drew me to Alacritty.
4. Alacritty still considers itself beta which means that they feel that performance can be improved. I can't say the same about their lack of features though.
5. Alacritty needs to be started as a new instance if I need more windows. This absolutely sucks. Tmux is not the answer to every use case.

This is where the advantages of Alacritty end for me.

Kitty:
1. The number of lines of code that are used to code Kitty or the language(s) that are used don't bother me. It is obvious this would be the case because Kitty has far more features than Alacritty (love the flexibility of the split layout and the tabs without the need for tmux BTW).
2. It is still more performant than Alacritty is even with more features and therefore more lines of code. The memory safety does bother me a bit but I leave Kitty running for months at a time and I don't see any significant memory bloat on my Mac.
3. Support for ligatures, tabs, windows is simply outstanding
4. I played around with various font settings and they don't seem to work.
5. I don't like that I have to do this TERM='xterm-256color' every time before I SSH into a remote host.

All-in-all I prefer Kitty to Alacritty mainly because of performance, ligatures, tabs and windows.

**Comment by ctrlcctrlv** [comment]
Just a few thoughts:

I know enough about fonts to say with absolute confidence that any issue you have with Kitty's fonts can be remedied, even issues the maintainer doesn't want to fix for understandable reasons—e.g. you do not need [`kitty-bitmap`](https://aur.archlinux.org/packages/kitty-bitmap/), bitmap fonts are possible in mainline kitty. (Cf. https://github.com/kovidgoyal/kitty/issues/97#issuecomment-579094656) Spaced out fonts are likely due to kitty's cell model—several workarounds exist, I'd need to see a render to comment further on that.

Kitty works on Windows.

Most of Kitty is actually written almost entirely in Python. The C code is very thin on the ground indeed, and most of it is macro based or just calling functions in other libraries (e.g. `kitty/fontconfig.c`). As the Rust compiler has [200 new memory safety issues discovered in past few weeks](https://www.infoq.com/news/2021/11/rudra-rust-safety/), your complaint seems unwarranted. Furthermore, Alacritty is just as vulnerable to memory issues in HarfBuzz, OpenGL bindings etc.

**Comment by dcow** [comment]
`kitty-terminfo` is in the Debian package repository now so you can just install that on hosts you are remotely accessing.

**Comment by ctrlcctrlv** [comment]
`community/kitty-terminfo` has been in Arch repos for quite a while too :)

**Comment by Luflosi** [comment]
And NixOS has `kitty.terminfo` :)

**Comment by kovidgoyal** [comment]
Yeah it was only a matter of time before kitty's terminfo became widely available in various distros. I am very happy to see the stranglehold of ncurses' maintainer on the terminal ecosystem being gradually broken.

**Comment by shezaan** [comment]
@dcow @ctrlcctrlv @Luflosi  Not everyone has root access to install packages on remote hosts, especially ones that aren't necessary when you are working in a corporate setting.

@ctrlcctrlv my comments are geared mainly towards the out-of-box experience with fonts in alacrity and for windows as well, although, I still can't figure out the fonts. The windows workaround seems pretty gnarly. As for Rust, I'm aware of the memory safety issues but they are still a lot fewer than any other language. That's why I didn't state that it is completely memory safe, also due to its dependence on the things that you mentioned none of which have been written in Rust.

In any case, I'm merely pointing out the user experience that I've faced. While workarounds do exist, they are just that... workarounds. If configs were the only things that needed tweaking, that would be understandable because that is considered part of the UX. Anything that needs to be done outside of tweaking the configs counts against the UX IMHO.
I really don't need to be sold on Kitty, I already am. I use it as my daily driver for home and work-related activities. I choose kitty after using Alacritty for over 2yrs. That should say a lot about how much I prefer kitty over alacritty. I merely stated what I did as a way to provide feedback on my experience which can perhaps be used to improve things if possible.
For what it's worth, I've attached an image of the 2 terminals so the difference in font rendering is hopefully visible. Kitty is on the left and Alacritty on the right.

![Screen Shot 2021-11-26 at 1 56 01 PM](https://user-images.githubusercontent.com/31293591/143622189-26ec6180-9542-4890-8c22-814b61c72c16.png)

**Comment by dcow** [comment]
@shezaan you don't need root to `kitty +kitten ssh you@host`. Do that once and the kitty terminfo is installed for your user (you can ssh normally afterwards).

**Comment by shezaan** [comment]
@dcow thanks I did not know that, are you saying that, that command will install the kitty.terminfo on the remote host right?

**Comment by Luflosi** [comment]
Yes, it creates the file `xterm-kitty` in `~/.terminfo/x/`.

**Comment by shezaan** [comment]
Thanks guys, this is very useful information but it doesn't seem to work when a jump box/bastion host is the proxy. It gets installed on the bastion, not the target it seems. It also doesn't work with some of the more corporate tools like scaleft (https://www.scaleft.com/) which uses a command like this... 
`sft ssh target-host`
I think I'll stick with my method which is to set `TERM='xterm-256color'` beforehand. It's simple albeit annoying but it does the job.

**Comment by Luflosi** [comment]
You could also try to manually create the two directories and copy the file manually. The file should be somewhere on your system, perhaps in `/usr/share/terminfo/x/`. If you can't find it on your system, it is also [in this repository](https://github.com/kovidgoyal/kitty/blob/master/terminfo/x/xterm-kitty).

**Comment by shezaan** [comment]
Unfortunately, it is not a viable option when there are 10's and for some others 100's of servers that have to be ssh'd into to copy `kitty.terminfo` even if it has to be done just once per server. I did try it on one and it does indeed work however it's not scalable especially when a tool like `sft` is involved because I would get asked my username and password during every ssh login for audit purposes.
Either ways, this is not a deal breaker for me. Kitty is indeed awesome and if there needs to be progress in the way of terminals then I would quote @kovidgoyal 
> If you want to perpetually stagnate the terminal ecosystem by insisting every terminal be xterm (the biggest boat anchor on the entire ecosystem), please just use xterm. kitty is not for you.

I could not have said it better.

**Comment by oblitum** [comment]
> e.g. you do not need [`kitty-bitmap`](https://aur.archlinux.org/packages/kitty-bitmap/), bitmap fonts are possible in mainline kitty. (Cf. [#97 (comment)](https://github.com/kovidgoyal/kitty/issues/97#issuecomment-579094656))

I'm sorry but that process is simply horrid, I invite everyone to not dig that and try it, just move on.

**Comment by ctrlcctrlv** [comment]
[`bitmapfont2otb`](https://github.com/ctrlcctrlv/bitmapfont2otb) automates a lot of the process, but the fontconfig parts remain relevant. If you need all the other features kitty offers but still want bitmap fonts for some reason, a “horrid" process is better than no process.

**Comment by oblitum** [comment]
@ctrlcctrlv agreed. Well, I'm a simple man and coming from Alacritty all I wished was undercurls (it's the sole thing I was missing there). As Alacritty doesn't have that and kitty doesn't have bitmap fonts out-of-the-box, I just web back to the good old terminal emulator I used to use ages ago, xfce4-terminal, it has everything I need (true colors, emojis and undercurls).

**Comment by Bananaman** [comment]
> As the Rust compiler has [200 new memory safety issues discovered in past few weeks](https://www.infoq.com/news/2021/11/rudra-rust-safety/), your complaint seems unwarranted.

Please don't lie.

You claim that 200 memory safety issues were found in the RUST COMPILER.

Nope. Completely false.

It was 264 memory safety issues in people's LIBRARIES, where they had used `unsafe {}` code blocks to BYPASS the Rust Compiler.

No more FUD from you, please.

**Comment by alexmozaidze** [comment]
I looked at issue 50 in Alacritty and it said at the top that it could be fixed with this PR https://github.com/alacritty/alacritty/pull/5696
I'm not sure, but I think this PR took inspiration from Kitty, I never used Kitty so I can't tell for sure.

**Comment by alexmozaidze** [comment]
Also, a word on Rust's perfectionism: I like it. Doing things the right way is good, if you look at a lot of crates in Rust you'll see that many of them are not perfectionist, they do the job but not always in the best and/or idiomatic way possible, which would make it harder to further manage the project. I am not a good dev, but I tried making some PR's, I made PR into rust_lisp crate in order to add configurability and BigInt support, and oh boi, clippy just wouldn't stop yelling at me! There was so much unnecessary code, and I think they even have a custom implementation of a linked list. Sure the library did its job, it executed Lisp, but the internals are a bit meesy. That's why I appreciate perfectionism in Rust - it makes reading projects easier, and my life less miserable when I want to make a PR that fixes my particular issue. Again, I'm dum, I'm not good at reading code.

**Comment by ctrlcctrlv** [comment]
@Bananaman What an overreaction. No more FUD from me? I mistyped—sorry, meant crates not compiler. My point still has some merit regarding security of Rust code with a lot of imports. Even so, the Rust compiler [is also not infallibly secure either](https://blog.rust-lang.org/2021/11/01/cve-2021-42574.html).

I maintain [a large Rust project](https://github.com/MFEK) which I founded. I have written more Rust code than 99% of developers. I use Rust/cargo every day. You seem to think I have some hatred of Rust/fanboyish ideas, but that is false. I will not pretend Rust is inherently more secure than it is though. Given Kitty is majority in Python, there is no reason at all to think that it has more memory safety issues than Alacritty.

@alexmozaidze Does @brundonsmith know you're over here shit talking his project? I doubt he'd appreciate it any more than I do. Lisp has `car` and `cdr`, it seems _entirely_ reasonable to have the `list::List` type I see in that crate. Furthermore, `clippy` doesn't "yell". It gives advisories and is not infallible, frequently making mistakes. That's by design. None of my projects use `clippy` for anything but occasional code-checking. (i.e. https://github.com/MFEK/glif/commit/c711dde78d553f477f86bd7e1b862434476681ac) Easily half of `clippy`'s advisories I considered for that commit were invalid. It's too much burden and also ugly for me to place clippy's ignore tags everywhere, so I won't be changing this, and you shouldn't expect other Rust developers to either.

**Comment by alexmozaidze** [comment]
I'm not talking shit on rust_lisp, I like this crate, I am pointing out some things I didn't like in this crate, and I think it's fine, as long as a program does its job - it's good, the only thing I don't like right now is the fact that it's a custom implementation of a linked list when std [already has one](https://doc.rust-lang.org/std/collections/struct.LinkedList.html), it would be possible to make a newtype called List and have both lisp-specific methods and methods provided by the wrapped `LinkedList` without reinventing the wheel. When I was making a PR Brundon said that his editor wasn't showing some of the lints I used, I suspect that his editor changed what lints are on by default, maybe? Idk.
Clippy without any additional configuration often gives valid warnings, I've encountered a false-positive only once with complexity lint, other times all the lints that clippy showed were pretty much valid. I agree on the fact that Rust is not entirely secure, it's a relatively young project, (and also the fact that no software is entirely secure, there are just vulnerabilities we don't know about yet) Rust mainly offers safety for stuff like memory without having a GC, you can cause a memory leak in safe Rust if you really wanted to. I don't want this discussion to get heated, just know that I use Rust mainly because of its type system more than anything else :P

**Comment by shaunsingh** [comment]
>I have written more Rust code than 99% of developers.

I would say any rust user has written more rust code than 99% of developers. Most developers haven't touched rust before

**Comment by kovidgoyal** [comment]
Please take discussions of rust elsewhere. Not all of us want rusty brains.

**Comment by mizzunet** [comment]
But Kitty seems to a brief but noticeable sluggish on the opening compared to Alacritty. Why that could be?

I'had been using Kitty for a while, then came across Alacritty. Now, I'm using Alacritty

**Comment by trygveaa** [comment]
> But Kitty seems to a brief but noticeable sluggish on the opening compared to Alacritty. Why that could be?

The slow startup time for kitty is mainly caused by two things as far as I know:
1. It has to initialize the GPU
2. It has to load the Python interpreter

Alacritty also has to do 1, so if kitty is slower than alacritty, I think it's likely due to 2.

However, kitty has a `--single-instance` option, which makes new terminal windows use the same process if kitty is already running. This makes new windows much faster to open since it avoids both 1 and 2, and since alacritty doesn't have anything like this, opening new windows in kitty (after the first one) with this option should be much faster than opening alacritty windows.

**Comment by page-down** [comment]
@trygveaa 
Unfortunately, I believe that `--single-instance` is handled in Python.

For the GPU part, you can choose to use iGPU, which might improve the startup speed.
https://sw.kovidgoyal.net/kitty/faq/#why-does-kitty-sometimes-start-slowly-on-my-linux-system

**Comment by trygveaa** [comment]
> @trygveaa Unfortunately, I believe that `--single-instance` is handled in Python.

Oh, okay. Though, even so kitty with `--since-instance` starts faster than alacritty at least on my machines.

> For the GPU part, you can choose to use iGPU, which might improve the startup speed. https://sw.kovidgoyal.net/kitty/faq/#why-does-kitty-sometimes-start-slowly-on-my-linux-system

Yeah, but I was thinking about general GPU initialization, not the specific issue mentioned there.

**Comment by kynrai** [comment]
I actually came back to have a look at kitty vs alacrity today and stumbled across this issue,

For me I was comparing idle CPU time and wide colour gamut support. I use Mac and the native Terminal.app and the popular iTerm2 both do not support the screens full colour gamut, kitty does not either and had higher idle CPU, alacrity had lower but not as low as native mac app but the difference in colour was instantly noticeable so i picked alacrity

**Comment by ttys3** [comment]
here comes an Alacritty user. I switched from Alacritty to Kitty today.

anyway, Kitty is just amazing !

**Comment by ShadiestGoat** [comment]
Font ligatures & start times. I came from urxvt wanting for something more modern - and in comparison to urxvt, both terminals have an extremely slow startup time, however, kitty less so. Alacritty has an open issue for font ligature support, but little work is being done on that. 
Another note of font, `╰' <- this is rendered much better on kitty than other terminals, and I don't understand _why_, since all use the same font.
I personally don't use the tabs & a lot of features of kitty, I have a tiled window manager which I like to use better than something like tmux. I would love it if this was modular, and you could just... not load this if you don't need it! But, as it stands I like kitty's performance.

**Comment by rizzini** [comment]
> So you're just here to troll then and not improve Kitty?

To be honest, this issue became a circlejerk thing rather than anything else.

**Comment by horta** [comment]
I still use Kitty on a daily basis. I use alacritty from time to time together with Kitty. They are both amazing tools for terminal people.

**Comment by rihaq** [comment]
2 reasons: 
Kitty not cross-platform 
Dev hate sixel

**Comment by ctrlcctrlv** [comment]
> > So you're just here to troll then and not improve Kitty?
> 
> To be honest, this issue became a circlejerk thing rather than anything else. No one is here to improve Kitty anymore.

<img src=https://user-images.githubusercontent.com/838783/189466020-838afe85-4614-416a-bdea-35744ead32ba.png width=200>

Speak for yourself, several of my changes resulted from comments made here.

For example:

>Dev hate sixel

Not true. It's partially my fault @kovidgoyal hasn't added Sixel support yet. He's right we can't add the broken C libsixel, and I haven't had time to write a libsixel C ABI compatible Rust libsixel we can link to instead. Kovid isn't against Sixel per se, unless his opinion changed since we last talked about it which I highly doubt, you're probably just being hyperbolic.

**Comment by ctrlcctrlv** [comment]
@ShadiestGoat 
>  Another note of font, `╰' <- this is rendered much better on kitty than other terminals, and I don't understand why, since all use the same font.

Because `kitty` has built-in to its GLSL programs code to draw those, they are not even dropped down to the font level. This is the right way to do it, and I don't see a way to disable it, so you can add one if you want (personally I see no need).

**Comment by ShadiestGoat** [comment]
@ctrlcctrlv 

> @ShadiestGoat
> 
> > Another note of font, `╰' <- this is rendered much better on kitty than other terminals, and I don't understand why, since all use the same font.
> 
> Because `kitty` has built-in to its GLSL programs code to draw those, they are not even dropped down to the font level. This is the right way to do it, and I don't see a way to disable it, so you can add one if you want (personally I see no need).

No, I too don't see why one would want to remove it. I quite like it. I did say it renders it better haha

**Comment by rihaq** [comment]
It wouldn't be hard to port kitty to Windows (don't ask me to do I don't know 'C' of computer)

**Comment by razamd784** [comment]
> Alacritty is a gimmick TE. That's why. Its authors are good at marketing. Its main gimmick is its alleged speed. Meanwhile we're really trying to build a terminal emulator useful for most people, we don't need to make ridiculous claims, which, if true, are only so due to severe lack of useful features.
> 
> I wrote a blog post a while ago about why I don't like Alacritty or indeed most of the Rust ecosystem: https://gist.github.com/ctrlcctrlv/978b3ee4f55d4b4ec415a985e01cb1c9

it's your opinion! I have tried both in my case alacritty is slightly faster and more smooth

**Comment by ctrlcctrlv** [comment]
It is, indeed, my opinion. "More smooth" is not a benchmark.

**Comment by razamd784** [comment]
> Kitty is super fast. Very stable. Well maintained. Provide the minimal features (split screen, tabs) for a fully working terminal (no need to add another indirection layer, like tmux, that will add complexity to your workflow, or hope that your OS have a window manager that will fill the features gap left by Alacritty).
> 
> I understand the point of Alacritty, their reasons, and it is a great terminal but it seems that Kitty has all of Alacritty's strenghts plus do not lack the minimal set of features to make it a fully working terminal out-of-the-box.

ghanta! The truth is you are the developer and just promoting kitty for your sake...Alacritty is best Minimal and highly customizable

**Comment by ctrlcctrlv** [comment]
@kovidgoyal Yesterday I asked you to unlock an issue, today I ask you to lock one. This issue has become nothing but a timesink when fanboys come here to defend their preferred TE.

**Comment by razamd784** [comment]
> @kovidgoyal Yesterday I asked you to unlock an issue, today I ask you to lock one. This issue has become nothing but a timesink when fanboys come here to defend their preferred TE.

First of all i didn't talk to you.... you are more worried than developer themselves? 🤣 what's your problem fanboy? second, You said you Don't care what i use or other...visit best neurologist near you for treatment as soon as possible....third, look at yourself first a biased kitty fanboy....fourth how you people have so much useless time to fight online? And yeah close this useless thing....if people need to choose one they try both and choose what is best for their need

**Comment by ctrlcctrlv** [comment]
lol

![image](https://user-images.githubusercontent.com/838783/203160993-dfc99653-708f-4631-afb5-c18cfdf8e7cc.png)

**Comment by razamd784** [comment]
> lol
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/838783/203160993-dfc99653-708f-4631-afb5-c18cfdf8e7cc.png)
> 
> 
Lol 😂this is my real account go to sleep you can't do anything 🤣 biased person ![image](https://user-images.githubusercontent.com/29495505/203162569-a85c9e21-156e-423a-818a-9ec5a19a0783.jpeg)

**Comment by SolitudeSF** [comment]
> it's your opinion! I have tried both in my case alacritty is slightly faster and more smooth

 it's your opinion! I have tried both in my case kitty is slightly faster and more smooth

**Comment by razamd784** [comment]
> > it's your opinion! I have tried both in my case alacritty is slightly faster and more smooth
> 
> 
> 
>  it's your opinion! I have tried both in my case kitty is slightly faster and more smooth
> 
> 

Ok thats what i am saying people should try both and find what they like Dont be @ctrlcctrlv

**Comment by SolitudeSF** [comment]
no, you just bought into alacritty propaganda and present your subjective optinion backed up by nothing as a sign of its superiority

**Comment by razamd784** [comment]
> no, you just bought into alacritty propaganda and present your subjective optinion backed up by nothing as a sign of its superiority

Okay but who start the kitty propaganda? And what's your backed up sign? Be a real man

**Comment by SolitudeSF** [comment]
"but you, but you". grow up.

**Comment by ctrlcctrlv** [comment]
You are both misunderstanding what I was using this issue for.

When people would write complaints with Kitty here, I'd investigate them. I'd fix the ones I could fix, or at least come to conclusions as to what is blocking the ones I could not (yet) fix.

Therefore, when I would ask people followup questions or start issues off of this one, that is what I was doing. My work on Kitty is not done to make anyone use, or stop using, any Terminal Emulator.

As stated, despite downvotes (what is this, Reddit?), I'll restate it:

> **I don't care what terminal you or anyone else uses dude.**

This issue no longer serves any purpose to the project because it has just become arguing about which terminal is better, and _ad hominem_&dagger; attacks.

Therefore I requested the issue be closed, as it serves no continued purpose to the project.

<sub>† _I.e., ad ursam viridem._</sub>
